viewing_party/party.py

```diff
@@ -1,23 +1,178 @@
+            movie = user_data["watchlist"][index]
+            print(user_data["watchlist"][index])
+    if move_movie:
+        (user_data["watched"]).append(movie)
+        del user_data["watchlist"][movie_index]
+    return user_data
+
+
+# 1st function in Wave 2
+def get_watched_avg_rating(user_data):
+
+    movie_count = 0
+    ratings_sum = 0
+    ratings_avg = 0
+
+    if user_data["watched"]:
+
+        for index in range(len(user_data["watched"])):
+
+            movie_count += 1
+            ratings_sum += user_data["watched"][index]["rating"]
+        ratings_avg = ratings_sum/movie_count
+    return ratings_avg
+
+
+# 2nd function in Wave 2
+def get_most_watched_genre(user_data):
+
+    most_watched_genre = user_data.get("watched",[])
+    
+    if not most_watched_genre: 
+        return None
+
+    genres = {}
+    for movie in most_watched_genre:
+        genre = movie.get("genre")
+        if genre in genres:
+            genres[genre] += 1
+        else:
+            genres[genre] = 1
+
+    most_watched_genre = max(genres, key=genres.get)
+
+    return most_watched_genre 
+
+
+# 1st function in Wave 3
+def get_unique_watched(user_data):
+
+    friends_watched_list = []
+    friends_watched_set = None
+    friends_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie["title"])
+            friends_watched_set = set(friends_watched_list)
+
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched_set and movie not in friends_list:
+            friends_list.append(movie)
+
+    return friends_list
+
+
+# 2nd function in Wave 3
+def get_friends_unique_watched(user_data):
+
+    user_watched = set([movie["title"] for movie in user_data["watched"]])
+    friends_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched and movie not in friends_list:
+                friends_list.append(movie)
+
+    return friends_list
+
+
+# 1st function in Wave 4
+def get_available_recs(user_data):
+
+    unseen_by_user = get_friends_unique_watched(user_data)
+    user_subscriptions = user_data["subscriptions"]
+    available_recs = []
+
+    for movie in unseen_by_user:
+        if movie["host"] in user_subscriptions:
+            available_recs.append(movie)
+    return available_recs
+
+
+#1st function Wave 5
+def get_new_rec_by_genre(user_data):
+
+    user_watched = set([movie["title"] for movie in user_data["watched"]])
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    genre_count = {}
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        genre_count[genre] = genre_count.get(genre, 0) + 1
+    best_genre = None
+    best_genre_count = 0
+    for genre, count in genre_count.items():
+        if count > best_genre_count:
+            best_genre = genre
+            best_genre_count = count

```

Comment:
This is a re-implementation of a function you've alreaady made! :O :P :D Consider this code, where you invoke the other function and get to delete ten lines! And the tests still pass :)

```python
def get_new_rec_by_genre(user_data):

    user_watched = set([movie["title"] for movie in user_data["watched"]])
    friends_unique_watched = get_friends_unique_watched(user_data)

    best_genre = get_most_watched_genre(user_data)

    recommended_movies = []
    # ...
```